### PR TITLE
PR: Enable Type-1 RNN support

### LINKED ISSUE #4930: [feature request] Multi-layer bidirectional RNN
Hello 
I request another type of multi-layer bidirectional RNN.
Currently forward output and backward output is concatenated after each layer.
But, for language modeling, we need independent forward rnn and backward rnn util the last layer and output concatenation is only need at the last layer.

thank you.

### PR DESCRIPTION
**Pull request by andfoy**
This PR introduces Type-1 behaviour on existing RNN implementations. For more information about the actual implementation of this PR, see https://github.com/pytorch/pytorch/issues/4930#issuecomment-589304807.

Fixes #4930

## Pending Tasks
- [x] CPU implementation (Unpacked)
- [x] CPU implementation (Packed)
- [x] CUDNN implemenatation (Unpacked)
- [x] CUDNN implementation (Packed)
- [x] MiOpen implementation (Unpacked)
- [x] MiOpen implementation (Packed)
- [x] Update Python interface
- [x] Tests

### CODE REVIEW

**Comment by rgommers** [comment]
@andfoy two test failures look real:
```
Mar 12 20:49:04 ======================================================================
Mar 12 20:49:04 ERROR [0.051s]: test_rnn_initial_hidden_state (__main__.TestNN)
Mar 12 20:49:04 ----------------------------------------------------------------------
Mar 12 20:49:04 Traceback (most recent call last):
Mar 12 20:49:04   File "test_nn.py", line 5551, in test_rnn_initial_hidden_state
Mar 12 20:49:04     output1, hidden1 = rnn(input, hidden)
Mar 12 20:49:04   File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 550, in __call__
Mar 12 20:49:04     result = self.forward(*input, **kwargs)
Mar 12 20:49:04   File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/rnn.py", line 794, in forward
Mar 12 20:49:04     self.cat_layer_fwd_bwd_states)
Mar 12 20:49:04 RuntimeError: The size of tensor a (32) must match the size of tensor b (10) at non-singleton dimension 0
Mar 12 20:49:04 
Mar 12 20:49:04 ======================================================================
Mar 12 20:49:04 ERROR [0.131s]: test_rnn_retain_variables_cpu_float64 (__main__.TestNNDeviceTypeCPU)
Mar 12 20:49:04 ----------------------------------------------------------------------
Mar 12 20:49:04 Traceback (most recent call last):
Mar 12 20:49:04   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 224, in instantiated_test
Mar 12 20:49:04     result = test(self, device_arg, dtype)
Mar 12 20:49:04   File "test_nn.py", line 9685, in test_rnn_retain_variables
Mar 12 20:49:04     self._test_rnn_retain_variables(device, dtype)
Mar 12 20:49:04   File "test_nn.py", line 9672, in _test_rnn_retain_variables
Mar 12 20:49:04     output = rnn(input)
Mar 12 20:49:04   File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 550, in __call__
Mar 12 20:49:04     result = self.forward(*input, **kwargs)
Mar 12 20:49:04   File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/rnn.py", line 794, in forward
Mar 12 20:49:04     self.cat_layer_fwd_bwd_states)
Mar 12 20:49:04 RuntimeError: The size of tensor a (6) must match the size of tensor b (5) at non-singleton dimension 0
Mar 12 20:49:04 
```

Also some merge conflicts to resolve.

**Comment by andfoy** [comment]
@rgommers @ezyang I'm having some segfaults on some ONNX tests, do you know who can give me a hand on this one?

**Comment by ezyang** [comment]
cc @houseroad 

But since these look like segfaults, I'd recommend "Rerun with SSH" and then gdb in the failing process (you'll need to exec into the docker image first)

**Comment by andfoy** [comment]
@ezyang, I think that type-2 bidirectional RNN spec is currently available on ONNX, so I prevented tracing RNNs when type-2 is specified.

**Comment by rgommers** [comment]
Hi @jamesr66a, looks like you're working on RNNs as well right now, so a couple of questions if you don't mind. 

First, this PR has run into major merge conflicts a couple of times, most recently due to your PR gh-35787 merged earlier today. So want to check in if there's more PRs like that in the works, and if so what's the best strategy to get this one merged?

Second, looks like you'd be a good reviewer for this PR since you're working on this code right now. Would you mind reviewing this one?

**Comment by jamesr66a** [comment]
Hi @rgommers, I have https://github.com/pytorch/pytorch/pull/34140 as well which should be landing pretty soon. That's the last of the `RNN.cpp` PRs coming on my end.

Other recent open PRs that affect RNNs include:
- https://github.com/pytorch/pytorch/pull/37159 @raghuramank100 
- https://github.com/pytorch/pytorch/pull/31934 @z-a-f 

Not sure if those are landing soon, respective authors would be able to weigh in there.

Lastly, I'll add myself as a reviewer and take a look when I get a chance!

File: aten/src/ATen/native/RNN.cpp [hunk 1]
```diff
@@ -1958,20 +2570,20 @@ static auto registry =
...
                 .kernel<decltype(quantized_gru_data), quantized_gru_data>(
                     DispatchKey::CPUTensorId))
-        .op("aten::quantized_gru.input_legacy(Tensor input, Tensor hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, bool batch_first) -> (Tensor, Tensor)",
+        .op("aten::quantized_gru.input_legacy(Tensor input, Tensor hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, bool type_2, bool batch_first) -> (Tensor, Tensor)",
             torch::RegisterOperators::options()
                 .kernel<
                     decltype(quantized_gru_input_legacy),
                     quantized_gru_input_legacy>(DispatchKey::CPUTensorId))
-        .op("aten::quantized_gru.data_legacy(Tensor data, Tensor batch_sizes, Tensor hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional) -> (Tensor, Tensor)",
+        .op("aten::quantized_gru.data_legacy(Tensor data, Tensor batch_sizes, Tensor hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, bool type_2) -> (Tensor, Tensor)",
```
Comment (jamesr66a) [inline aten/src/ATen/native/RNN.cpp:2586]: It's probably better to add the `type_2` flag as an optional kwarg with default to maintain backward-compatibility

**Review by jamesr66a — COMMENTED** [review]
Are there any tests for this?

File: aten/src/ATen/native/RNN.cpp (hunk 1 above)
Comment (andfoy) [reply aten/src/ATen/native/RNN.cpp:2586]: Sorry for the newbie question, is there any example of adding a kwarg on C10?

**Comment by andfoy** [comment]
@jamesr66a Thanks for the review! I'm adding the tests later on today.

File: torch/nn/modules/rnn.py [hunk 2]
```diff
@@ -24,11 +24,13 @@ def apply_permutation(tensor, permutation, dim=1):
 
 class RNNBase(Module):
     __constants__ = ['mode', 'input_size', 'hidden_size', 'num_layers', 'bias',
-                     'batch_first', 'dropout', 'bidirectional']
+                     'batch_first', 'dropout', 'bidirectional',
+                     'cat_layer_fwd_bwd_states']
```
Comment (rgommers) [inline torch/nn/modules/rnn.py:28]: This is a very descriptive but quite awkward name, would be good to shorten it. If it's really a boolean option (there can never be >2 methods, even in the future), then perhaps `concat=True`. I noticed `tf.keras.layers.Bidirectional` uses `merge_mode='concat'` (and has >2 options).

File: test/test_nn.py [hunk 3]
```diff
@@ -5526,8 +5526,8 @@ def compare_cpu_gpu(outputs_cpu, outputs_gpu):
                     self.assertEqual(cpu_weight.grad.data, gpu_weight.grad.data, atol=5e-5)
 
         for module in (nn.RNN, nn.LSTM, nn.GRU):
-            for bias, bidirectional, batch_first, contig, variable_len, lens_as_tensor \
-                    in product((True, False), repeat=6):
+            for bias, bidirectional, type_2, batch_first, contig, variable_len, lens_as_tensor \
```
Comment (rgommers) [inline test/test_nn.py:5529]: This test check CPU vs GPU and also covers ROCM, so all code is exercised. However it just checks for consistency between CPU and GPU implementations, not correctness (as far as I can tell). There are several other tests for `nn.RNN` & co with `bidirectional=True` that you could run for type1 and type2.

File: torch/nn/modules/rnn.py [hunk 4]
```diff
@@ -321,6 +372,11 @@ class RNN(RNNBase):
             RNN layer except the last layer, with dropout probability equal to
             :attr:`dropout`. Default: 0
         bidirectional: If ``True``, becomes a bidirectional RNN. Default: ``False``
+        cat_layer_fwd_bwd_states: If ``True``, then the forward and backward
+            hidden states will be concatenated and fed as input to each
+            layer, except for the first one. If set to ``False``, each hidden
+            state will be processed independently and concatenated just at the
+            end of the network. Default: ``True``
```
Comment (rgommers) [inline torch/nn/modules/rnn.py:379]: Should say something like ```This option is ignored when `bidirectional=True`.```

**Comment by rgommers** [comment]
There's a relevant failure in job `pytorch_windows_vs2019_py36_cpu_test2
` `test_quantized_rnn` and 2 failures `test_RNN_cpu_vs_cudnn_with/no_dropout` in job `pytorch_windows_vs2019_py36_cuda10.1_test2`.

**Comment by ezyang** [comment]
@andfoy just wanted to check in how this was doing, and if you are blocked at all.

**Comment by andfoy** [comment]
Hi @ezyang, thanks for asking. Right now I'm adding the test for testing the implementation of Type-1, the one that compares CPU Vs GPU outputs is now working. I expect that this test is the final change required before review

**Comment by andfoy** [comment]
@rgommers @ezyang I've just added the individual Type-1 test, thus I think that this PR is finally ready for review

**Comment by ezyang** [comment]
@jamesr66a do you think you'd be able to do the final review on this?

File: torch/nn/modules/rnn.py (hunk 2 above)
Comment (rgommers) [reply torch/nn/modules/rnn.py:28]: @andfoy any thoughts on this? I suspect other modes could be added in the future, so `merge_mode` would probably make sense here.

File: test/test_nn.py (hunk 3 above)
Comment (rgommers) [reply test/test_nn.py:5587]: test coverage looks much better now, thanks

**Comment by rgommers** [comment]
Still a couple of CI failures that are real.

File: torch/nn/modules/rnn.py (hunk 2 above)
Comment (andfoy) [reply torch/nn/modules/rnn.py]: > so merge_mode would probably make sense here.

I think this is an adequate solution, but I don't know if we should include it as part of this PR, or leave it for a refactoring PR after this one?

Comment (rgommers) [reply torch/nn/modules/rnn.py]: After the rename to `concat=True` this API looks fine. Maybe `merge_mode` is still an improvement, but not sure. Changing to string/enum here would be a bit of work, so let's see if anyone else has an opinion.

File: aten/src/ATen/native/RNN.cpp (hunk 1 above)
Comment (rgommers) [reply aten/src/ATen/native/RNN.cpp:2583]: This seems like the likely cause for:
```
The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not. 
Broken ops: [
    aten::rnn_relu.input(Tensor input, Tensor hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, bool batch_first) -> (Tensor, Tensor)
...
```

I don't know what the answer is to how to do this. @jamesr66a or @ezyang any suggestion?

Comment (jamesr66a) [reply aten/src/ATen/native/RNN.cpp:2583]: Hey, it should be as simple as using `=False` in the schema string. Here's an example: https://github.com/pytorch/pytorch/blob/d8b9448c621143feaa4eea5bf20e5c586b1ffe8f/aten/src/ATen/native/quantized/library.cpp#L85

**Comment by rgommers** [comment]
@andfoy I suggest to take care of the backwards compatibility test (adding `=False` to not add a required arg to the C++ API) first. Getting access to a ROCm cloud machine is still a work-in-progress.

**Comment by ezyang** [comment]
@andfoy if you're blocked on ROCm, just disable this feature on ROCm and let the AMD crew take a look at it.

**Comment by andfoy** [comment]
> just disable this feature on ROCm and let the AMD crew take a look at it.

Do I add some sort of warning/error?

**Comment by ezyang** [comment]
Yes, make it error, and disable the test on ROCm

**Comment by andfoy** [comment]
@ezyang @rgommers This one is finally ready, but I have two unrelated build issues on Jenkins

**Comment by rgommers** [comment]
> I have two unrelated build issues on Jenkins

Those indeed look unrelated, and https://ezyang.github.io/pytorch-ci-hud/build/pytorch-master says the CentOS job is still broken.

**Comment by rgommers** [comment]
All my previous comments were addressed. This looks ready for a detailed review @jamesr66a. 

I've also asked on the issue with the feature request if people who wanted this feature would like to test.

**Comment by ezyang** [comment]
@jamesr66a will you be able to review this?

**Pull request closed without merging**